File: web_version/pmugg_files/node/nodeStats.js
ms.nodeStats = function() {
	this.nodes = {};
	this.count = {};
	for (var i = 0; i < ms.nodeStats.nodeTypes.length; i++) {
		var name = ms.nodeStats.nodeTypes[i];
		this.nodes[name] = {};
		this.count[name] = 0;
	}
	this.changeList = [];
	this.costChange = {};
	this.model = null;
	this.resetCostChange();
	this.badCollection = new ms.collection(this, 'vertex');
	this.unloopedCollection = new ms.collection(this, 'face');
	this.badCollection.getNode().save();
	this.unloopedCollection.getNode().save();
};

ms.nodeStats.nodeTypes = [
	'cell', 'collection', 'endpoint', 'face', 'faceConnection', 'faceGroup', 'line', 'lineSegment', 'lineState', 'vertex', 'vertexState',
	'ringInstance', 'bspNode', 'bspEdge', 'bspPolygon',
];

ms.nodeStats.costTerms = ['lineDistance', 'valence', 'reject'];

ms.nodeStats.prototype.setModel = function(model) {
	this.model = model;
};

ms.nodeStats.prototype.getModel = function(node) {
	return this.model;
}

ms.nodeStats.prototype.addNode = function(node) {
	var name = node.getName();
	this.nodes[name][node.getId()] = node;
	this.count[name]++;
};

ms.nodeStats.prototype.removeNode = function(node) {
	var name = node.getName();
	delete this.nodes[name][node.getId()];
	this.count[name]--;
};

ms.nodeStats.prototype.addVertex = function(node) {
	var vertex = node.element;
	if (vertex.hasConflict()) {
		node.connect(this.badCollection);
	} else {
		if (node.has('collection') && node.get('collection')) {
			node.disconnect(this.badCollection);
		}
	}
};

ms.nodeStats.prototype.removeVertex = function(node) {
	if (node.has('collection') && node.get('collection')) {
		node.disconnect(this.badCollection);
	}
};

ms.nodeStats.prototype.updateFace = function(face, unlooped) {
	var node = face.getNode();
	if (unlooped) {
		node.connect(this.unloopedCollection);
	} else {
		if (node.has('collection') && node.get('collection')) {
			node.disconnect(this.unloopedCollection);
		}
	}
};

ms.nodeStats.prototype.getBadVertices = function() {
	return this.badCollection.getVertices();
};

ms.nodeStats.prototype.getUnloopedFaces = function() {
	return this.unloopedCollection.getFaces();
};

ms.nodeStats.prototype.getElements = function(name) {
	if (!this.nodes.hasOwnProperty(name)) {
		return [];
	}
	var elements = [];
	return Object.values(this.nodes[name]).map(function(node) {
		return node.getElement();
	});
};

ms.nodeStats.prototype.getCount = function(name) {
	return this.count.hasOwnProperty(name) ? this.count[name] : 0
};

ms.nodeStats.prototype.addToChangeList = function(node) {
	this.changeList.push(node);
};

ms.nodeStats.prototype.apply = function(func) {
	for (var type in this.nodes) {
		for (var id in this.nodes[type]) {
			func(this.nodes[type][id]);
		}
	}
};

ms.nodeStats.prototype.save = function() {
	this.changeList.forEach(function(node) {
		node.save();
	});
	this.changeList = [];
	this.resetCostChange();
};

ms.nodeStats.prototype.restore = function() {
	this.changeList.forEach(function(node) {
		node.restoreDestroyed();
	});
	this.changeList.forEach(function(node) {
		node.restoreCreated();
	});
	this.changeList = [];
	this.resetCostChange();
};

ms.nodeStats.get = function() {
	return window.controller.currentController.synthesizer.mutator.nodeStats;
};
	
// This only used for testing to verify nothing weird has happened.
ms.nodeStats.verifyAll = function() {
	ms.nodeStats.get().verifyAll();
};

ms.nodeStats.prototype.verifyAll = function() {
	for (var name in this.nodes) {
		for (var id in this.nodes[name]) {
			var node = this.nodes[name][id];
			if (node.isDestroyed()) {
				window.console.log('Destroyed node');
				debugger;
			}
		}
	}
	
	for (var id in this.nodes['lineState']) {
		var element = this.nodes['lineState'][id].element;
		if (!element.cell.activeStates.default.includes(element)) {
			window.console.log('Cell missing node.');
			debugger;
		}
	}
};

ms.nodeStats.prototype.componentCount = function() {
	var count = 0;
	var faces = this.nodes['face'];
	for (id in faces) {
		var face = faces[id];
		if (face.getElement().isHole()) {
			count++;
		}
	}
	return count;
};

ms.nodeStats.prototype.resetCostChange = function() {
	var self = this;
	ms.nodeStats.costTerms.forEach(function(costTerm) {
		self.costChange[costTerm] = 0;
	});
	if (ms.optimizer.detailedCost) {
		this.costChange.details = {};
	}
};

File: cpp_version/pmugg/node/node_stats.cpp
#include "node_stats.h"
#include "model.h"
#include <stdexcept>

namespace ms {

const std::vector<std::string> NodeStats::nodeTypes = {
    "cell", "collection", "endpoint", "face", "faceConnection", "faceGroup",
    "line", "lineSegment", "lineState", "vertex", "vertexState",
    "ringInstance", "bspNode", "bspEdge", "bspPolygon"
};

const std::vector<std::string> NodeStats::costTerms = {
    "lineDistance", "valence", "reject"
};

NodeStats::NodeStats()
    : model(nullptr)
    , badCollection(this, "vertex")
    , unloopedCollection(this, "face")
{
    for (const auto& name : nodeTypes) {
        nodes[name] = std::map<int, Node*>();
        count[name] = 0;
    }
    resetCostChange();
    badCollection.getNode()->save();
    unloopedCollection.getNode()->save();
}

NodeStats::~NodeStats() {
    // Clean up any remaining nodes
    for (auto& [_, typeNodes] : nodes) {
        for (auto& [_, node] : typeNodes) {
            delete node;
        }
    }
}

void NodeStats::setModel(Model* m) {
    model = m;
}

Model* NodeStats::getModel() const {
    return model;
}

void NodeStats::addNode(Node* node) {
    auto name = node->getName();
    nodes[name][node->getId()] = node;
    count[name]++;
}

void NodeStats::removeNode(Node* node) {
    auto name = node->getName();
    nodes[name].erase(node->getId());
    count[name]--;
}

void NodeStats::addVertex(Node* node) {
    auto vertex = static_cast<Vertex*>(node->getElement());
    if (vertex->hasConflict()) {
        node->connect(&badCollection);
    } else {
        if (node->has("collection") && node->get<Collection*>("collection")) {
            node->disconnect(&badCollection);
        }
    }
}

void NodeStats::removeVertex(Node* node) {
    if (node->has("collection") && node->get<Collection*>("collection")) {
        node->disconnect(&badCollection);
    }
}

void NodeStats::updateFace(Face* face, bool unlooped) {
    auto node = face->getNode();
    if (unlooped) {
        node->connect(&unloopedCollection);
    } else {
        if (node->has("collection") && node->get<Collection*>("collection")) {
            node->disconnect(&unloopedCollection);
        }
    }
}

std::vector<Vertex*> NodeStats::getBadVertices() const {
    return badCollection.getVertices();
}

std::vector<Face*> NodeStats::getUnloopedFaces() const {
    return unloopedCollection.getFaces();
}

std::vector<void*> NodeStats::getElements(const std::string& name) const {
    if (!nodes.contains(name)) {
        return {};
    }
    std::vector<void*> elements;
    elements.reserve(nodes.at(name).size());
    for (const auto& [_, node] : nodes.at(name)) {
        elements.push_back(node->getElement());
    }
    return elements;
}

int NodeStats::getCount(const std::string& name) const {
    return count.contains(name) ? count.at(name) : 0;
}

void NodeStats::addToChangeList(Node* node) {
    changeList.push_back(node);
}

void NodeStats::apply(std::function<void(Node*)> func) {
    for (const auto& [_, typeNodes] : nodes) {
        for (const auto& [_, node] : typeNodes) {
            func(node);
        }
    }
}

void NodeStats::save() {
    for (auto* node : changeList) {
        node->save();
    }
    changeList.clear();
    resetCostChange();
}

void NodeStats::restore() {
    for (auto* node : changeList) {
        node->restoreDestroyed();
    }
    for (auto* node : changeList) {
        node->restoreCreated();
    }
    changeList.clear();
    resetCostChange();
}

void NodeStats::resetCostChange() {
    for (const auto& term : costTerms) {
        costChange[term] = 0;
    }
    // TODO: Handle detailed cost if needed
}

void NodeStats::verifyAll() {
    for (const auto& [name, typeNodes] : nodes) {
        for (const auto& [_, node] : typeNodes) {
            if (node->isDestroyed()) {
                throw std::runtime_error("Destroyed node found");
            }
        }
    }

    for (const auto& [_, node] : nodes["lineState"]) {
        auto element = static_cast<LineState*>(node->getElement());
        if (!element->cell->activeStates["default"].contains(element)) {
            throw std::runtime_error("Cell missing node");
        }
    }
}

int NodeStats::componentCount() const {
    int count = 0;
    for (const auto& [_, node] : nodes.at("face")) {
        auto face = static_cast<Face*>(node->getElement());
        if (face->isHole()) {
            count++;
        }
    }
    return count;
}

NodeStats* NodeStats::get() {
    return Controller::getCurrentController()->getSynthesizer()->getMutator()->getNodeStats();
}

void NodeStats::verifyAll_() {
    get()->verifyAll();
}

} // namespace ms 